kafka-persistence/src/main/scala/com/evolutiongaming/kafka/flow/kafkapersistence/KafkaPersistence.scala

```diff
@@ -0,0 +1,169 @@
+    val zero: BytesByKey = Map.empty
+  }
+
+  private case class MissingOffsetError(topicPartition: TopicPartition) extends NoStackTrace
+
+  final case class Of[F[_], K, S, A](create: Partition => F[KafkaPersistence[F, K, S, A]])
+
+  object Of {
+    def apply[F[_] : BracketThrowable : Producer : ToBytes[*[_], S] : FromBytes[*[_], S] : FromTry : Log : Sync, S, A](
+                                                                                                                        consumerOf: ConsumerOf[F],
+                                                                                                                        consumerConfig: ConsumerConfig,
+                                                                                                                        stateTopic: Topic
+                                                                                                                      ): Of[F, KafkaKey, S, A] =
+      this { partition =>
+        for {
+          stateData <- readState(
+            consumerOf = consumerOf,
+            consumerConfig = consumerConfig,
+            stateTopic = stateTopic,
+            partition = partition
+          )
+          stateRef <- Ref.of(stateData)
+        } yield KafkaPersistence[F, S, A](stateTopic, stateRef.stateInstance, Ref.of(none[Snapshot[S]]).map(_.stateInstance))
+      }
+  }
+
+  def apply[F[_] : Producer : Monad : FromTry : Log,
+    S: FromBytes[F, *] : ToBytes[F, *],
+    A](
+        stateTopic: Topic,
+        monadState: MonadState[F, BytesByKey],
+        buffers: F[MonadState[F, Option[Snapshot[S]]]]
+      ): KafkaPersistence[F, KafkaKey, S, A] = {
+    val snapshotsOf = new SnapshotsOf[F, KafkaKey, S] {
+      override def apply(key: KafkaKey): F[Snapshots[F, S]] =
+        for {
+          buffer <- buffers
+        } yield Snapshots(
+          key,
+          new SnapshotDatabase[F, KafkaKey, S] {
+            override def get(key: KafkaKey) =
+              for {
+                state <- monadState.get
+                s <- state.get(key.key).traverse(bytes => FromBytes[F, S].apply(bytes.toArray, stateTopic))
+              } yield s
+
+            override def persist(key: KafkaKey, snapshot: S) = produce(snapshot.some)
+
+            override def delete(key: KafkaKey) = produce(none)
+
+            private def produce(snapshot: Option[S]) =
+              Producer[F]
+                .send(new ProducerRecord(topic = stateTopic, key = key.key.some, value = snapshot))
+                .flatten
+                .void
+          },
+          buffer
+        )
+    }
+    val keysOf = new KeysOf[F, KafkaKey] {
+      override def apply(key: KafkaKey) = Keys.empty
+
+      override def all(applicationId: String, groupId: String, topicPartition: TopicPartition) =
+        Stream.lift(monadState.get
+          .map(bytes => Stream.from[F, List, KafkaKey](bytes.keys.map(KafkaKey(applicationId, groupId, topicPartition, _)).toList))).flatten
```

Comment:
This line seems to be a bit intensive? Is it worth to split it to several pieces or lines?

Reply:
Agree, moved stream creation logic to a helper method.